--- cpuinfo-8a1772a0c5c447df2d18edf33ec4603a8c9c04a6/tools/auxv-dump.c	2025-11-12 11:43:14.476909303 +0800
+++ cpuinfo-8a1772a0c5c447df2d18edf33ec4603a8c9c04a6/tools/auxv-dump.c	2025-11-12 11:44:00.350010006 +0800
@@ -3,7 +3,14 @@
 
 #include <dlfcn.h>
 #include <errno.h>
+#if CPUINFO_ARCH_ARM64 ||                                                     \
+        CPUINFO_ARCH_ARM && defined(__GLIBC__) && defined(__GLIBC_MINOR__) && \
+                (__GLIBC__ > 2 || __GLIBC__ == 2 && __GLIBC_MINOR__ >= 16)
 #include <sys/auxv.h>
+#else
+#define AT_HWCAP 16
+#define AT_HWCAP2 26
+#endif
 
 #include <cpuinfo.h>
 
