From 495bb81da96b75a0a2c800fcc61f1c32f52c9114 Mon Sep 17 00:00:00 2001
From: "Dave.Yu im294" <Dave.Yu@augentix.com>
Date: Mon, 9 Jun 2025 13:51:52 +0800
Subject: [PATCH] Avoid registering RAND algorithm

---
 src/meson.build |  1 -
 src/provider.c  | 12 ------------
 2 files changed, 13 deletions(-)

diff --git a/src/meson.build b/src/meson.build
index 11aacc8..b12f8a1 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -11,7 +11,6 @@ pkcs11_provider_sources = [
   'interface.c',
   'objects.c',
   'provider.c',
-  'random.c',
   'session.c',
   'signature.c',
   'slot.c',
diff --git a/src/provider.c b/src/provider.c
index 20a1984..a99e4ea 100644
--- a/src/provider.c
+++ b/src/provider.c
@@ -49,7 +49,6 @@ struct p11prov_ctx {
 
     OSSL_ALGORITHM *op_digest;
     OSSL_ALGORITHM *op_kdf;
-    OSSL_ALGORITHM *op_random;
     OSSL_ALGORITHM *op_exchange;
     OSSL_ALGORITHM *op_signature;
     OSSL_ALGORITHM *op_asym_cipher;
@@ -536,7 +535,6 @@ static void p11prov_ctx_free(P11PROV_CTX *ctx)
 
     OPENSSL_free(ctx->op_digest);
     OPENSSL_free(ctx->op_kdf);
-    OPENSSL_free(ctx->op_random);
     OPENSSL_free(ctx->op_exchange);
     OPENSSL_free(ctx->op_signature);
     OPENSSL_free(ctx->op_asym_cipher);
@@ -1083,13 +1081,6 @@ static CK_RV operations_init(P11PROV_CTX *ctx)
     TERM_ALGO(signature);
     TERM_ALGO(asym_cipher);
 
-    /* handle random */
-    ret = p11prov_check_random(ctx);
-    if (ret == CKR_OK) {
-        ADD_ALGO_EXT(RAND, random, prop, p11prov_rand_functions);
-        TERM_ALGO(random);
-    }
-
     return CKR_OK;
 }
 
@@ -1238,9 +1229,6 @@ p11prov_query_operation(void *provctx, int operation_id, int *no_cache)
     case OSSL_OP_KDF:
         *no_cache = ctx->status == P11PROV_UNINITIALIZED ? 1 : 0;
         return ctx->op_kdf;
-    case OSSL_OP_RAND:
-        *no_cache = ctx->status == P11PROV_UNINITIALIZED ? 1 : 0;
-        return ctx->op_random;
     case OSSL_OP_KEYMGMT:
         *no_cache = 0;
         return ctx->op_keymgmt;
-- 
2.47.1

