--- a/Rules.mak
+++ b/Rules.mak
@@ -607,7 +607,7 @@
 # Collect all CFLAGS components
 CFLAGS := $(XWARNINGS) $(CPU_CFLAGS) $(SSP_CFLAGS) \
 	-nostdinc -I$(top_builddir)include \
-	-I$(top_srcdir)include -include libc-symbols.h \
+	-I$(top_srcdir)include -include uclibc-yocto-bootstrap.h -include libc-symbols.h \
 	-I$(top_srcdir)libc/sysdeps/linux/$(TARGET_ARCH) \
 	-I$(top_srcdir)libc/sysdeps/linux \
 	-I$(top_srcdir)ldso/ldso/$(TARGET_ARCH) \
